File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import defaultdict,deque
def solution(begin, target, words):
    
    def one_diff(w1,w2):
        count=0
        for i in range(len(w1)):
            if w1[i]!=w2[i]:
                count+=1
        return count==1
    
    # INIT GRAPH
    N=len(words)
    graph=defaultdict(list)
    for u in range(N):
        for v in range(N):
            if one_diff(words[u],words[v]):
                graph[u].append(v)
         
    # BFS (Multi-Source)
    visited=[False]*N
    distance=[0]*N
    queue=deque()
    for i in range(N):
        if one_diff(begin,words[i]):
            queue.append(i)
            visited[i]=True
            distance[i]=1
    
    while queue:
        cur_node=queue.popleft()
        # base-condition
        if words[cur_node]==target:
            return(distance[cur_node])
        for next_node in graph[cur_node]:
            if not visited[next_node]:
                visited[next_node]=True
                distance[next_node]=distance[cur_node]+1
                queue.append(next_node)
                
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    public int solution(String begin, String target, String[] words) {
        
        boolean[] visited=new boolean[words.length];
        Deque<String[]> queue=new ArrayDeque<>();
        queue.addLast(new String[]{begin,"0"});
        
        while(!queue.isEmpty()){
            String[] cur=queue.pollFirst();
            String cur_state=cur[0];
            int depth=Integer.parseInt(cur[1]);
            if (cur_state.equals(target)) return depth;
            for(int i=0;i<words.length;i++){
                if(oneDiff(cur_state,words[i])&&!visited[i]){
                    visited[i]=true;
                    queue.addLast(new String[]{words[i],String.valueOf(depth+1)});
                }
            }
        }
        return 0;
    }
    
    private boolean oneDiff(String a,String b){
        int count=0;
        for (int i=0;i<a.length();i++){
            if(a.charAt(i)!=b.charAt(i)){
                count+=1;
            }
        }
        return count==1;
    }
}